src/PIL/ImageFont.py

```diff
@@ -634,6 +634,7 @@ def getmask2(
         should have mode ``RGBA``. Otherwise, it should have mode ``1``.
 
         :param text: Text to render.
+        :param fill: A fill function.
```

Comment:
It's not clear to me what is the use for this parameter. It defaults to a function in the private interface, yet AFAICT it is not passed a different value anywhere within Pillow. My best guess is that it was supposed to be used with `ImageDraw.setfill`, deprecated before the fork and removed in Pillow 3.0.0. I think it was undocumented because it was inteded to be private? This docstring is also not particularly helpful.

Would it make more sense to deprecate it instead?

Reply:
Yeah, it probably would be a good idea to deprecate it.

The definition:

```python
    def getmask2(
        self,
        text,
        mode="",
        fill=Image.core.fill,
        direction=None,
        features=None,
        language=None,
        stroke_width=0,
        anchor=None,
        ink=0,
        *args,
        **kwargs,
    ):
```

The tricky thing is that it's the 3rd of many parameters. We can remove it in Pillow 10, a breaking change, but we would want people to be able to modify their code during the 9.x deprecation period and have it still work in 10 with no further changes.

We're not setting `fill` and so use keyword params for all those after `fill`:

```python
        return self.getmask2(
            text,
            mode,
            direction=direction,
            features=features,
            language=language,
            stroke_width=stroke_width,
            anchor=anchor,
            ink=ink,
        )[0]
```

And:

```python
                mask, offset = font.getmask2(
                    text,
                    mode,
                    direction=direction,
                    features=features,
                    language=language,
                    stroke_width=stroke_width,
                    anchor=anchor,
                    ink=ink,
                    *args,
                    **kwargs,
                )
```

So our code won't break. 

We don't know how others are calling `getmask2`. They're _probably_ not setting `fill` but is it safe to assume so?

Well, maybe in Pillow 10 we can set those later arguments to be [keyword-only arguments](https://peps.python.org/pep-3102/).

If they're not using `fill` they must be using the later ones as keyword-only arguments. 

* No warning. 
* They don't need to change their code.

If they are using `fill`, they could be using later ones as either positional or keyword-only.

* In any case, we warn `fill` is deprecated and to use keyword-only for the later ones.
* They can change their code during the deprecation period and it'll still work when the Pillow 10 breaking change happens.

Reply:
I don't think it is necessary to make them keyword-only, and this decision can be deferred until the parameter is actually removed anyway.

* If they aren't using `fill`, they are passing them as keyword args and don't need to change their code.
* If they are using `fill` (I expect it is very uncommon), they will have to change to keyword args until it is removed whether or not they are keyword-only.

After the parameter is removed, both can change to use positional args moving forward.

However, it might be worth making them keyword-only due to how many there are. I remember I struggled with keeping them backwards-compatible when working on some previous PRs (e.g. #4930). In that case, it would be worth deprecating positional args for all of the font-related functions.

Reply:
Ah yes, good point. 👍 

> However, it might be worth making them keyword-only due to how many there are. I remember I struggled with keeping them backwards-compatible when working on some previous PRs (e.g. https://github.com/python-pillow/Pillow/pull/4930). 

Yeah, some of them have grown a lot of arguments over the years!

> In that case, it would be worth deprecating positional args for all of the font-related functions.

If I'm calling `getmask2("my text", "RGBA")` now, how would we detect that to warn?

And not warn for `getmask2(text="my text", mode="RGBA")`?

Reply:
I would keep `text` as the only positional argument since it is mandatory.

To detect whether `mode` is passed as positional, the function can be changed to `getmask2(text, _mode=None, fill=Image.core.fill, ..., *args, mode="", **kwargs)`. If `_mode` is set, it was passed as a positional argument and should raise a warning. Otherwise it was passed as a keyword argumnent.

Reply:
Yes, I see.

Would you like to put together a PR to deprecate `fill` from `getmask2`? It's pretty clear that parameter isn't much use. 

And for "deprecating positional args for all of the font-related functions": it'd be nice to see a draft PR for a single function first, to get an idea, in case we decide not to do it and avoid wasted effort. Mention which other functions you'd like to be included. No need to do all the docs and stuff, and maybe just a partial test. Feel free to open an issue for further discussion :)

Reply:
Sure, but I would prefer to wait until #6184 is merged/rejected first :)

Reply:
#6184 has now been merged.

Reply:
> Would you like to put together a PR to deprecate fill from getmask2? It's pretty clear that parameter isn't much use.

See #6220.

> And for "deprecating positional args for all of the font-related functions"

Will do a bit later.